fdbserver/BlobMigrator.actor.cpp

```diff
@@ -30,54 +30,302 @@
+		state Transaction tr(self->db_);
+		loop {
+			tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			try {
+				state RangeResult serverList = wait(tr.getRange(serverListKeys, CLIENT_KNOBS->TOO_MANY));
+				ASSERT(!serverList.more && serverList.size() < CLIENT_KNOBS->TOO_MANY);
+				for (auto& server : serverList) {
+					state UID id = decodeServerListValue(server.value).id();
+					RangeResult ranges = wait(krmGetRanges(&tr, serverKeysPrefixFor(id), keys));
+					bool owning = false;
+					for (auto& r : ranges) {
+						if (r.value == serverKeysTrue) {
+							owning = true;
+							break;
+						}
+					}
+					if (owning) {
+						dprint("Unassign {} from storage server {}\n", keys.toString(), id.toString());
+						wait(krmSetRange(&tr, serverKeysPrefixFor(id), keys, serverKeysFalse));
+					}
+				}
+				wait(tr.commit());
+				return Void();
+			} catch (Error& e) {
+				wait(tr.onError(e));
+			}
+		}
+	}
+
+	// Main server loop
+	ACTOR static Future<Void> serverLoop(Reference<BlobMigrator> self) {
+		self->actors_.add(waitFailureServer(self->interf_.ssi.waitFailure.getFuture()));
+		self->actors_.add(handleRequest(self));
+		self->actors_.add(handleUnsupportedRequest(self));
 		loop {
-			choose {
-				when(HaltBlobMigratorRequest req = waitNext(self->blobMigratorInterf.haltBlobMigrator.getFuture())) {
-					req.reply.send(Void());
-					TraceEvent("BlobMigratorHalted", self->blobMigratorInterf.id()).detail("ReqID", req.requesterID);
-					break;
+			try {
+				choose {
+					when(HaltBlobMigratorRequest req = waitNext(self->interf_.haltBlobMigrator.getFuture())) {
+						req.reply.send(Void());
+						TraceEvent("BlobMigratorHalted", self->interf_.id()).detail("ReqID", req.requesterID);
+						break;
+					}
+					when(wait(self->actors_.getResult())) {}
 				}
-				when(wait(self->actors.getResult())) {}
+			} catch (Error& e) {
+				dprint("Unexpected serverLoop error {}\n", e.what());
+				throw;
 			}
 		}
 		return Void();
 	}
 
+	// Handle StorageServerInterface APIs
+	ACTOR static Future<Void> handleRequest(Reference<BlobMigrator> self) {
+		state StorageServerInterface ssi = self->interf_.ssi;
+		loop {
+			try {
+				choose {
+					when(GetShardStateRequest req = waitNext(ssi.getShardState.getFuture())) {
+						dprint("Handle GetShardStateRequest\n");
+						Version version = maxVersion(self);
+						GetShardStateReply rep(version, version);
+						req.reply.send(rep); // return empty shards
+					}
+					when(WaitMetricsRequest req = waitNext(ssi.waitMetrics.getFuture())) {
+						// dprint("Handle WaitMetricsRequest\n");
+						state WaitMetricsRequest waitMetricsRequest = req;
+						// FIXME get rid of this delay. it's a temp solution to avoid starvaion scheduling of DD
+						// processes
+						wait(delay(1));
```

Comment:
even though this delay is temporary, in general it is bad to have a delay in a large choose when, because it blocks the entire loop from processing new messages. Its better to add an actor to an actor collection with will wait the 1 second delay then send the reply

Reply:
fixed. thank you!